cpp/include/Ice/IncomingRequest.h

```diff
@@ -0,0 +1,71 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+#ifndef ICE_INCOMING_REQUEST_H
+#define ICE_INCOMING_REQUEST_H
+
+#include "ConnectionF.h"
+#include "Current.h"
+#include "ObjectAdapterF.h"
+
+namespace Ice
+{
+    class InputStream;
+
+    /**
+     * Represent a request received by a connection. It's the argument to the dispatch function on Object.
+     * @remarks IncomingRequest is neither copyable nor movable. It can be used only on the dispatch thread.
+     * @see Object::dispatch
+     * \headerfile Ice/Ice.h
+     */
+    class ICE_API IncomingRequest final
+    {
+    public:
+        /**
+         * Construct an IncomingRequest object.
+         * @param requestId The request ID. It's 0 for oneway requests.
+         * @param connection The connection that received the request. It's null for collocated invocations.
+         * @param adapter The object adapter to set in Current.
+         * @param inputStream The input stream buffer over the incoming Ice protocol request message. The stream is
+         * positioned at the beginning of the request header - the next data to read is the identity of the target.
+         */
+        IncomingRequest(
+            int32_t requestId,
+            ConnectionPtr connection,
+            ObjectAdapterPtr adapter,
+            InputStream& inputStream);
+
+        IncomingRequest(const IncomingRequest&) = delete;
+        IncomingRequest(IncomingRequest&&) noexcept = delete;
+        IncomingRequest& operator=(const IncomingRequest&) = delete;
+        IncomingRequest& operator=(IncomingRequest&&) = delete;
+
+        /**
+         * Return the current object of the request.
+         */
+        Current& current() noexcept { return _current; }
```

Comment:
This allows a middleware to modify current. Add a comment to speculate on what a middleware can change?

It could add or remove an entry in current.context? 

The equivalent API in IceRPC is roughly:
https://github.com/icerpc/icerpc-csharp/blob/a1b10df451c97fd0ef21049da939dab5268818af/src/IceRpc/IncomingRequest.cs#L15

We don't have a comment explaining why you may want to change the features.